common/djangoapps/lang_pref/middleware.py

```diff
@@ -21,20 +22,33 @@ def process_request(self, request):
         """
         If a user's UserPreference contains a language preference, use the user's preference.
         """
+        languages = released_languages()
+        system_released_languages = [seq[0] for seq in languages]
+        preferred_language = request.META.get('HTTP_ACCEPT_LANGUAGE', '')
+        lang_headers = [seq[0] for seq in parse_accept_lang_header(preferred_language)]
+
         # If the user is logged in, check for their language preference
         if request.user.is_authenticated():
             # Get the user's language preference
             user_pref = get_user_preference(request.user, LANGUAGE_KEY)
+            language = settings.LANGUAGE_CODE
             # Set it to the LANGUAGE_SESSION_KEY (Django-specific session setting governing language pref)
             if user_pref:
-                request.session[LANGUAGE_SESSION_KEY] = user_pref
-            else:
-                # Setting the session language to the browser language, if it is supported.
-                preferred_language = request.META.get('HTTP_ACCEPT_LANGUAGE', '')
-                lang_headers = [seq[0] for seq in parse_accept_lang_header(preferred_language)]
-                languages = released_languages()
-                for browser_lang in lang_headers:
-                    if browser_lang in [seq[0] for seq in languages]:
-                        if request.session.get(LANGUAGE_SESSION_KEY, None) != browser_lang:
-                            request.session[LANGUAGE_SESSION_KEY] = unicode(browser_lang)
-                        break
+                if user_pref in system_released_languages:
+                    language = user_pref
+                else:
+                    # update the user language preference as system is not supporting the existing preferred language.
+                    # check support of browser languages, otherwise set the system default language.
+                    for browser_lang in lang_headers:
+                        if browser_lang in system_released_languages:
+                            language = unicode(browser_lang)
+                            break
+                    set_user_preference(request.user, LANGUAGE_KEY, language)
```

Comment:
I thing yes, I had a conversation with @mattdrayer 

I was thinking it could be possible that a language might be removed from the list of "supported languages" for some reason -- perhaps the translations become corrupted, or another operator chooses to remove a language for some other valid reason.  In this case the user's preference would no longer be valid.
So we could then consider the browser language setting, if it is supported by the system.
And if that is also unsupported, we could fall back onto the system default language